File: plugin/src/shared/sc/plugin2020/Field.java
package sc.plugin2020;

import java.util.Stack;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamAsAttribute;
import com.thoughtworks.xstream.annotations.XStreamConverter;
import com.thoughtworks.xstream.annotations.XStreamImplicit;
import com.thoughtworks.xstream.converters.collections.ArrayConverter;
import com.thoughtworks.xstream.converters.collections.CollectionConverter;
import com.thoughtworks.xstream.converters.extended.ToAttributedValueConverter;
import sc.api.plugins.IField;
import sc.plugin2020.util.CubeCoordinates;
import sc.shared.PlayerColor;

@XStreamAlias("field")
public class Field implements IField {

  @XStreamImplicit
  private Stack<Piece> pieces = new Stack<Piece>();
  @XStreamAsAttribute
  private boolean obstructed = false;

  // NOTE that we use primitives instead of CubeCoordinates to let XStream searialize coordinates as attributes of field tag
  @XStreamAsAttribute
  private int x;
  @XStreamAsAttribute
  private int y;
  @XStreamAsAttribute
  private int z;

  public Field(CubeCoordinates position){
    this.x = position.x; this.y = position.y; this.z = position.z;
  }

  public Field(CubeCoordinates position, boolean obstructed){
    this.x = position.x; this.y = position.y; this.z = position.z;
    this.obstructed = obstructed;
  }

  public Field(CubeCoordinates position, Stack<Piece> pieces){
    this.x = position.x; this.y = position.y; this.z = position.z;
    this.pieces = pieces;
  }

  public FieldState getFieldState(){
    if (obstructed)
      return FieldState.OBSTRUCTED;

    if (!pieces.isEmpty()) {

      if(pieces.peek().getOwner() == PlayerColor.RED)
        return FieldState.RED;

      if(pieces.peek().getOwner() == PlayerColor.BLUE)
        return FieldState.BLUE;
    }
    return FieldState.EMPTY;
  }

  public boolean isObstructed(){
    return obstructed;
  }

  public void setObstructed(boolean o){ obstructed = o;}

  public Stack<Piece> getPieces(){
    return(pieces);
  }

  public CubeCoordinates getPosition(){
    return new CubeCoordinates(this.x, this.y, this.z);
  }

  public int getX() {
    return getPosition().x;
  }
}

File: plugin/src/shared/sc/plugin2020/Field.kt
package sc.plugin2020

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamAsAttribute
import com.thoughtworks.xstream.annotations.XStreamImplicit
import sc.api.plugins.IField
import sc.plugin2020.util.CubeCoordinates
import sc.shared.PlayerColor
import java.util.*

@XStreamAlias("field")
class Field(
        x: Int = 0,
        y: Int = 0,
        z: Int = -x - y,
        pieces: Stack<Piece> = Stack(),
        @XStreamAsAttribute
        val isObstructed: Boolean = false
): CubeCoordinates(x, y, z), IField {
    
    // Custom getter because an empty collection in XML gets deserialized to null, not to an empty collection!
    // See http://x-stream.github.io/faq.html#Serialization_implicit_null
    @XStreamImplicit
    var pieces = pieces
        get() {
            @Suppress("SENSELESS_COMPARISON")
            if(field == null) field = Stack()
            return field
        }
    
    val fieldState: FieldState
        get() {
            if(isObstructed)
                return FieldState.OBSTRUCTED
            return when(owner) {
                PlayerColor.RED -> FieldState.RED
                PlayerColor.BLUE -> FieldState.BLUE
                null -> FieldState.EMPTY
            }
        }
    
    val isEmpty: Boolean
        get() = pieces.isEmpty() && !isObstructed
    
    val hasOwner: Boolean
        get() = !pieces.isEmpty() && !isObstructed
    
    val coordinates: CubeCoordinates
        get() = CubeCoordinates(x, y, z)
    
    val owner: PlayerColor?
        get() = if(pieces.isEmpty()) null else pieces.peek().owner
    
    constructor(position: CubeCoordinates, obstructed: Boolean): this(position.x, position.y, position.z, isObstructed = obstructed)
    
    constructor(position: CubeCoordinates, vararg pieces: Piece): this(position.x, position.y, position.z, pieces.toCollection(Stack()))
    
    constructor (x: Int, y: Int, obstructed: Boolean): this(x, y, isObstructed = obstructed)
    
    constructor(x: Int, y: Int, vararg pieces: Piece): this(x, y, pieces = pieces.toCollection(Stack()) as Stack<Piece>)
    
    constructor(field: Field): this(field.x, field.y, field.z, field.pieces.toCollection(Stack()), field.isObstructed)
    
    override fun equals(other: Any?): Boolean {
        if(this === other) return true
        if(javaClass != other?.javaClass)
            return if(other is CubeCoordinates) super.equals(other) else false
        
        other as Field
        
        if(x != other.x) return false
        if(y != other.y) return false
        if(z != other.z) return false
        if(pieces != other.pieces) return false
        if(isObstructed != other.isObstructed) return false
        
        return true
    }
    
    override fun hashCode(): Int {
        var result = x
        result = 31 * result + y
        result = 31 * result + z
        result = 31 * result + pieces.hashCode()
        result = 31 * result + isObstructed.hashCode()
        return result
    }
    
}
